### ISSUE TITLE
[Bug]Backend Inconsistency: Major Loss (1.04) and Gradient mismatch in recurrent WDCG topologies

### ISSUE THREAD

**Issue by Beanlycool** [keras-team/keras#22697]
I have identified a critical numerical inconsistency between tensorflow and torch backends in Keras 3. For a specific Weighted Directed Cyclic Graph (WDCG) topology involving concurrent spatial and recurrent paths, the Loss Difference is $1.041992 \times 10^0$ and the Max Relative Gradient Difference is 1.50\%.

[test1.py](https://github.com/user-attachments/files/26852550/test1.py)

Output: \nLoss Difference: 1.041992e+00
Param Grad 2 | Max Abs Diff: 3.729150e+01 | Rel Diff: 1.50%

System information:
OS: [Ubuntu 22.04 / Linux]
Python version: [3.11]
Keras version: [3.14.0]
TensorFlow version: [2.16.1]
PyTorch version: [2.5.1]

**Comment by jeffcarp** [comment]
Hi, running in Colab I get a smaller difference:
`Param Grad 2 | Max Abs Diff: 7.812500e-03 | Rel Diff: 0.00%`

Can you upgrade the versions of above and re-run?

**Comment by ArjunSrivastava1** [comment]
hey there @jeffcarp , i too ran the same tests in my own systems and colab both and got this:
> Loss Difference: 1.953125e-03 |Param Grad 2 | Max Abs Diff: 4.089355e-03 | Rel Diff: 0.00%

as similar results is confirmed for all colab environments, and my own local systems too

thus i suggest for this issue to be closed as and when you can

**Comment by jeffcarp** [comment]
@ArjunSrivastava1 can you share your exact Colab repro?

**Comment by ArjunSrivastava1** [comment]
sure, um, i deleted the tab when i did the repro cus its been a few days honeslty, but here u are with another one i just ran:

input = cell code above, take the code in the issue and put it in a new colab env you have, even existing notebooks

runtime configs - CPU
the output is produced as such:

> \nLoss Difference: 6.835938e-03
> Param Grad 2 | Max Abs Diff: 6.347656e-03 | Rel Diff: 0.00%
runtime = 47s, 

> Focus the last run cell
> 12:29 XX (0 minutes ago)
> executed in 47.282s

runtime config - T4 GPU
the output is produced as such:

> \nLoss Difference: 1.953125e-03
> Param Grad 2 | Max Abs Diff: 3.540039e-03 | Rel Diff: 0.00%
runtime = 34s

> Focus the last run cell
>12:32 XX (0 minutes ago)
> executed in 34.285s

standard runtime env used by me is T4 GPU, which also produced the results that i posted before, i hope this much is enough @jeffcarp

**Comment by jeffcarp** [comment]
Thanks, so it sounds like this is irrespective of accelerator (CPU/GPU).

Based on some digging, it looks like this could be caused by differences of 10e-7 to 10e-8 between the way TF and Torch compute things (which are normal and expected) which get compounded by the gradient calculation as it goes through your recurrent model architecture.

Here's a repro that doesn't involve Keras at all:

```python
import numpy as np
import tensorflow as tf
import torch

# 1. Initialize identical inputs and weights in float32
np.random.seed(42)
# A 64x64 matrix designed to have a spectral radius slightly > 1 to force compounding
x_np = np.random.normal(size=(1, 64)).astype(np.float32)
W_np = np.random.normal(scale=0.2, size=(64, 64)).astype(np.float32)

# 2. Setup Native TensorFlow
x_tf = tf.constant(x_np)
W_tf = tf.Variable(W_np)

# 3. Setup Native PyTorch
x_pt = torch.tensor(x_np)
W_pt = torch.tensor(W_np, requires_grad=True)

print("--- Forward Pass Differences (Step-by-Step) ---")
with tf.GradientTape() as tape:
    out_tf = x_tf
    out_pt = x_pt
    
    for step in range(100):
        # Native TF
        out_tf = tf.matmul(out_tf, W_tf)
        out_tf = tf.nn.tanh(out_tf)
        
        # Native PyTorch
        out_pt = torch.matmul(out_pt, W_pt)
        out_pt = torch.tanh(out_pt)
        
        # Track the divergence step-by-step
        diff = np.max(np.abs(out_tf.numpy() - out_pt.detach().numpy()))
        if step % 10 == 0:
          print(f"Step {step+1:02d} Max Abs Diff: {diff:.6e}")

    # Simple loss: sum of all elements
    loss_tf = tf.reduce_sum(out_tf)
    loss_pt = torch.sum(out_pt)

# 4. Compute Gradients (Backpropagation compounding)
grad_tf = tape.gradient(loss_tf, W_tf).numpy()
loss_pt.backward()
grad_pt = W_pt.grad.numpy()

print("\n--- Final Gradient Differences (Backprop) ---")
grad_diff = np.max(np.abs(grad_tf - grad_pt))
rel_diff = grad_diff / (np.mean(np.abs(grad_tf)) + 1e-8)

print(f"Max Abs Gradient Diff : {grad_diff:.6e}")
print(f"Relative Gradient Diff: {rel_diff:.2%}")
```

```
--- Forward Pass Differences (Step-by-Step) ---
Step 01 Max Abs Diff: 2.235174e-07
Step 11 Max Abs Diff: 6.556511e-07
Step 21 Max Abs Diff: 1.527369e-06
Step 31 Max Abs Diff: 1.274049e-06
Step 41 Max Abs Diff: 1.668930e-06
Step 51 Max Abs Diff: 4.038215e-06
Step 61 Max Abs Diff: 7.748604e-06
Step 71 Max Abs Diff: 9.149313e-06
Step 81 Max Abs Diff: 1.945347e-05
Step 91 Max Abs Diff: 8.232892e-06

--- Final Gradient Differences (Backprop) ---
Max Abs Gradient Diff : 2.502441e-03
Relative Gradient Diff: 0.02%
```

**Comment by ArjunSrivastava1** [comment]
@jeffcarp , i can confirm ur results, and the differing ways both the stacks work, so this really is just expected behavior
for the numpy code not involving keras, my repro comes out to be this:

T4
```
--- Forward Pass Differences (Step-by-Step) ---
Step 01 Max Abs Diff: 3.064051e-07
Step 11 Max Abs Diff: 1.311302e-06
Step 21 Max Abs Diff: 2.294779e-06
Step 31 Max Abs Diff: 2.965331e-06
Step 41 Max Abs Diff: 6.973743e-06
Step 51 Max Abs Diff: 4.321337e-06
Step 61 Max Abs Diff: 6.332994e-06
Step 71 Max Abs Diff: 7.390976e-06
Step 81 Max Abs Diff: 1.361221e-05
Step 91 Max Abs Diff: 6.482005e-06

--- Final Gradient Differences (Backprop) ---
Max Abs Gradient Diff : 2.929688e-03
Relative Gradient Diff: 0.02%
```

CPU
```
--- Forward Pass Differences (Step-by-Step) ---
Step 01 Max Abs Diff: 2.235174e-07
Step 11 Max Abs Diff: 6.556511e-07
Step 21 Max Abs Diff: 1.527369e-06
Step 31 Max Abs Diff: 1.274049e-06
Step 41 Max Abs Diff: 1.668930e-06
Step 51 Max Abs Diff: 4.038215e-06
Step 61 Max Abs Diff: 7.748604e-06
Step 71 Max Abs Diff: 9.149313e-06
Step 81 Max Abs Diff: 1.945347e-05
Step 91 Max Abs Diff: 8.232892e-06

--- Final Gradient Differences (Backprop) ---
Max Abs Gradient Diff : 2.502441e-03
Relative Gradient Diff: 0.02%
```

i say we've looked into everything enough to close the issue, and i recommend for closing it now unless you want to do some deeper digging on the differences or a few docs change? should i open a pr containing a broader note in docs regarding this expected difference
